src/main/java/org/truffleruby/core/array/ArrayWriteNormalizedNode.java

```diff
@@ -41,16 +41,49 @@ public Object writeWithin(DynamicObject array, int index, Object value,
         return value;
     }
 
+    // writing within an array after converting it to be mutable
+
+    @Specialization(guards = { "!strategy.isStorageMutable()",
+            "isInBounds(array, index)", "strategy.matches(array)", "mutableStrategy.accepts(value)"
```

Comment:
Guards order: mutable is always the most specific and should be last.

Reply:
Sadly this isn't compatible with making `accept()` throw `UnsupportedOperationException` for immutable cases. Shall we go for `matches`, `mutable`, and then `accepts` as our standard ordering?

Reply:
That sounds good.